src/ATen/Scalar.h

```diff
@@ -103,10 +102,8 @@ class Scalar {
   union {
     double d;
     int64_t i;
+    TensorImpl* t;
```

Comment:
I don't like doing this because it leads to subtle errors. 

For instance, the code as it stands needs to also override the assignment operator, copy-constructor, move constructor, and move constructors for scalar and implement them correctly. I'd rather not do this unless you know this was already a performance bottleneck.

Reply:
I'm trying to work around the circular dependency between Scalar.h and Tensor.h. The methods in Tensor.h requires the Scalar type defined so that default arguments like "Scalar alpha=1" work. Using TensorImpl* instead of Tensor was the easiest way to break the cycle. 

I'm open to other solutions. For example, we could refactor out the smart-pointer aspect of Tensor into a base class and use that instead of Tensor.

We could also implement the constructors and assignment operator. They aren't trivial, but they aren't too hard to get right.

Reply:
Short term: I think the `TensorBase` solution seems best for now.
Long Term: Because we are sorting out default argument order and confusing method overloads, it seems like we will soon be at a point when we can consider merging Scalar/Tensor into one type without bad side effects. When that happens this union behavior will move into Tensor in some form.

Reply:
I refactored out TensorBase from Tensor. It's not ideal either:
1) Some constructors and assignment operators have to be defined in both classes since they're not inherited due to C++ inheritance rules.
2) Scalar can end up backed by null tensors. (This was also true before this PR). For example:

```c++
Scalar s1(tensor);
Scalar s2 = std::move(s1);
// s1 is now backed by a null tensor
```